yarl/_quoting_c.pyx

```diff
@@ -91,7 +93,15 @@ cdef struct Writer:
 
 
 cdef inline void _init_writer(Writer* writer):
-    writer.buf = &BUFFER[0]
+    cdef char *buf
+    if IS_GIL_DISABLED:
+        buf = <char *> PyMem_Malloc(BUF_SIZE)
```

Comment:
Could the code always use a thread-local buffer?
In this case, the amortized time will be O(1), like with the static buffer.

Reply:
Implemented in [6be69fc](https://github.com/aio-libs/yarl/pull/1456/commits/6be69fc0797cc9fecc948b9d930d24418212d0de). Could you please have a look and let me know whether this approach seems okay?

Reply:
I may have missed the intent here, but I was expecting something like what was asked for in https://github.com/cython/cython/issues/6745

Reply:
Because the buffer is only ever needed within the lifetime of `_do_quote_or_skip`, making it a local to that function effectively makes it a thread-local as well. So we're getting the same benefit without the need for C++'s `thread_local`.